### ISSUE TITLE
Internal seed of EBM seems to be not fixed

### ISSUE THREAD

**Issue by superbunny38** [interpretml/interpret#424]
Hello, I am a user of Explainable Boosting Machine, and thanks to your effort I was trying to make an interpretation out of the survey data. However, it turned out that although I fixed the seed the scores of score boards are not reproducible. I get different score boards every time I build an EBM model with the same used seed with other settings fixed the same as well. The performance of the model (ex. f1-score, accuracy and etc.) however, is the same and thus reproducible. I was trying to visualize the scoreboard again in my colab, but I wasn't able to get the same scoreboard as before.

I'll wait for your response.
Thank you :)

**Comment by paulbkoch** [comment]
Hi @superbunny38 -- This is very surprising to me.  I routinely check code updates against previous versions using exact reproducibility.  Could you post a little of the code you are using when creating the EBMs?  Also, what version of interpret are you using?

**Comment by superbunny38** [comment]
Hello, because I had to use colab on my private data, I installed ebm library every time I started Runtime via command:

`!pip install interpret`

When training EBM, I used the code below:

`best_update = False
try:
  for _ in tqdm(range(70)):
    target_or = target
    num_cols = df[features]._get_numeric_data().columns
    categorical_cols = set(features)-set(list(num_cols))
    one_hot_encoded_data = pd.get_dummies(df[features], columns = list(categorical_cols))

    X,y = one_hot_encoded_data,target_or
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=_)
    X_resampled, y_resampled = SMOTE().fit_resample(X_train, y_train)
    smote_ebm = ExplainableBoostingClassifier(random_state=seed, n_jobs=-1)
    smote_ebm.fit(X_resampled, y_resampled) 
    smote_train_acc = smote_ebm.score(X_train,y_train)
    smote_test_acc = smote_ebm.score(X_test, y_test)
    smote_f1 = f1_score(list(y_test),smote_ebm.predict(X_test),average='weighted')
    if smote_f1< 0.8 and best_update == False:
      print(f">>early stopping for loop: {smote_f1}")
      break
    if smote_f1> best_f1 and smote_f1> 0.8:
      best_f1 = smote_f1
      best_update = True
    # print("smote_f1:",smote_f1)
    if smote_f1>0.94:
      raise GetOutOfLoop
  print("best:",best_f1)
except GetOutOfLoop:
    pass`

**Comment by superbunny38** [comment]
Hello, because I had to use colab on my private data, I installed ebm library every time I started Runtime via command:

`!pip install interpret`

When training EBM, I used the code below:

`best_update = False
try:

  for _ in tqdm(range(70)):
    # target_feature1,target_feature2 = best_f1_dict['used_features'][0],best_f1_dict['used_features'][1]11

    # target_and = np.logical_and(target1,target2)*1
    target_or = target
    num_cols = df[features]._get_numeric_data().columns
    categorical_cols = set(features)-set(list(num_cols))
    one_hot_encoded_data = pd.get_dummies(df[features], columns = list(categorical_cols))

    X,y = one_hot_encoded_data,target_or
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=_)
    X_resampled, y_resampled = SMOTE().fit_resample(X_train, y_train)
    smote_ebm = ExplainableBoostingClassifier(random_state=seed, n_jobs=-1)
    smote_ebm.fit(X_resampled, y_resampled) 
    smote_train_acc = smote_ebm.score(X_train,y_train)
    smote_test_acc = smote_ebm.score(X_test, y_test)
    smote_f1 = f1_score(list(y_test),smote_ebm.predict(X_test),average='weighted')
    if smote_f1< 0.8 and best_update == False:
      print(f">>early stopping for loop: {smote_f1}")
      break
    if smote_f1> best_f1 and smote_f1> 0.8:
      best_f1 = smote_f1
      best_update = True
    # print("smote_f1:",smote_f1)
    if smote_f1>0.94:
      raise GetOutOfLoop
  print("best:",best_f1)
except GetOutOfLoop:
    pass`

**Comment by superbunny38** [comment]
`target_or = target
num_cols = df[features]._get_numeric_data().columns
categorical_cols = set(features)-set(list(num_cols))
one_hot_encoded_data = pd.get_dummies(df[features], columns = list(categorical_cols))

X,y = one_hot_encoded_data,target_or
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=_)
X_resampled, y_resampled = SMOTE().fit_resample(X_train, y_train)
smote_ebm = ExplainableBoostingClassifier(random_state=seed, n_jobs=-1)
smote_ebm.fit(X_resampled, y_resampled) 
smote_train_acc = smote_ebm.score(X_train,y_train)
smote_test_acc = smote_ebm.score(X_test, y_test)
smote_f1 = f1_score(list(y_test),smote_ebm.predict(X_test),average='weighted')
if smote_f1< 0.8 and best_update == False:
  print(f">>early stopping for loop: {smote_f1}")
  break
if smote_f1> best_f1 and smote_f1> 0.8:
  best_f1 = smote_f1
  best_update = True
`

**Comment by paulbkoch** [comment]
Hi @superbunny38 -- can you try setting the random_state for train_test_split and also SMOTE?

**Comment by superbunny38** [comment]
@paulbkoch  Hello, I already fixed the seed for train_test_split as _ which denotes the iteration idx. However, thanks to your suggestion, it turns out that the seed for SMOTE is not fixed too. I will run the code with SMOTE fixed and then tell you if that was the sole problem.

Thank you! :D
